theano/gpuarray/ctc.py

```diff
@@ -0,0 +1,185 @@
+                raise RuntimeError('libwarpctc.so could not be found. ',
+                                   'Please check the config.ctc.root variable.')
+        return dirs
+
+    def c_libraries(self):
+        return ["warpctc", "gpuarray"]
+
+    def c_header_dirs(self):
+        dirs = [os.path.dirname(__file__), pygpu.get_include()]
+        if ctc_enabled:
+            # We assume here that the header is available at the include directory
+            # of the CTC root directory.
+            dirs.append(os.path.join(config.ctc.root, "include"))
+        return dirs
+
+    def c_headers(self):
+        return ['ctc.h', 'numpy_compat.h', 'gpuarray_helper.h', 'gpuarray/types.h',
+                'gpuarray_api.h', 'gpuarray/array.h', 'gpuarray/util.h', 'gpuarray/extension.h']
+
+    def get_params(self, node):
+        return node.inputs[0].type.context
+
+    def make_node(self, activations, labels, input_lengths):
+        if not ctc_enabled:
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'GpuConnectionistTemporalClassification Op '
+                               'can not be constructed.')
+
+        context = infer_context_name(activations, labels, input_lengths)
+
+        t_activations = as_gpuarray_variable(activations,
+                                             context_name=context)
+        # Ensure activations array is C-contiguous
+        t_activations = gpu_contiguous(t_activations)
+
+        # Labels and input lengths are always on the CPU
+        t_labels = T.as_tensor_variable(labels)
+        t_input_lengths = T.as_tensor_variable(input_lengths)
+
+        if t_activations.type.dtype != 'float32':
+            raise TypeError('Activations must use the float32 type!')
+
+        if t_labels.type.dtype != 'int32':
+            raise TypeError('Labels must use the int32 type!')
+
+        if t_input_lengths.type.dtype != 'int32':
+            raise TypeError('Label lengths must use the int32 type!')
+
+        # Return only the cost. Gradient will be returned by grad()
+        self.default_output = 0
+
+        costs = GpuArrayType(dtype='float32',
+                             broadcastable=(False,),
+                             context_name=context)()
+        outputs = [costs]
+
+        if self.compute_grad:
+            gradients = GpuArrayType(dtype='float32',
+                                     broadcastable=(False, False, False,),
+                                     context_name=context)()
+            outputs += [gradients]
+
+        return theano.Apply(self, inputs=[t_activations, t_labels, t_input_lengths],
+                            outputs=outputs)
+
+    def L_op(self, inputs, outputs, output_grads):
+        if not ctc_enabled:
+            raise RuntimeError('Baidu CTC is not enabled and '
+                               'GpuConnectionistTemporalClassification Op '
+                               'can not be constructed.')
```

Comment:
This check is not useful because if that function is called, we already have a GpuConnectionistTemporalClassification object.